pkg/controllers/dnspolicy/dnspolicy_healthchecks.go

```diff
@@ -22,91 +26,104 @@ type healthChecksConfig struct {
-	// Get the associated DNSRecord
-	dnsRecords, err := r.getDNSRecords(ctx, policy)
-	if err != nil {
-		return err
+func (r *DNSPolicyReconciler) reconcileHealthChecks(ctx context.Context, dnsPolicy *v1alpha1.DNSPolicy, gwDiffObj *reconcilers.GatewayDiff) error {
+	log := crlog.FromContext(ctx)
+
+	// Delete Health checks for each gateway no longer referred by this policy
+	for _, gw := range gwDiffObj.GatewaysWithInvalidPolicyRef {
+		log.V(1).Info("reconcileHealthChecks: gateway with invalid policy ref", "key", gw.Key())
+		_, err := r.reconcileGatewayHealthChecks(ctx, gw.Gateway, nil)
+		if err != nil {
+			return err
+		}
 	}
-	if len(dnsRecords) == 0 {
-		log.Info("No DNSRecord found matching policy", "policy", policy)
-		return nil
+
+	healthCheckConfig := getHealthChecksConfig(dnsPolicy)
+	var allResults []dns.HealthCheckResult
+
+	// Reconcile Health checks for each gateway directly referred by the policy (existing and new)
+	for _, gw := range append(gwDiffObj.GatewaysWithValidPolicyRef, gwDiffObj.GatewaysMissingPolicyRef...) {
+		log.V(1).Info("reconcileHealthChecks: gateway with valid and missing policy ref", "key", gw.Key())
+		results, err := r.reconcileGatewayHealthChecks(ctx, gw.Gateway, healthCheckConfig)
+		if err != nil {
+			return err
+		}
+		allResults = append(allResults, results...)
 	}
 
-	// Get the configuration for the health checks. If no configuration is
-	// set, ensure that the health checks are deleted
-	config := getHealthChecksConfig(policy)
+	return r.reconcileHealthCheckStatus(allResults, dnsPolicy)
+}
 
+func (r *DNSPolicyReconciler) reconcileGatewayHealthChecks(ctx context.Context, gateway *gatewayv1beta1.Gateway, config *healthChecksConfig) ([]dns.HealthCheckResult, error) {
 	var allResults []dns.HealthCheckResult
 
+	gatewayAccessor := traffic.NewGateway(gateway)
+	dnsRecords, err := r.HostService.GetDNSRecordsFor(ctx, gatewayAccessor)
+	if err != nil {
+		return allResults, err
+	}
+
 	for _, dnsRecord := range dnsRecords {
 		// Keep a copy of the DNSRecord to check if it needs to be updated after
 		// reconciling the health checks
 		dnsRecordOriginal := dnsRecord.DeepCopy()
 
-		var results []dns.HealthCheckResult
-
-		if config == nil {
-			log.V(3).Info("No health check for DNSPolicy, ensuring deletion")
-			results, err = r.reconcileHealthCheckDeletion(ctx, dnsRecord, policy)
-			if err != nil {
-				return err
-			}
-		} else {
-			log.Info("reconciling health checks")
-			results, err = r.reconcileHealthChecks(ctx, log, dnsRecord, policy, config)
-			if err != nil {
-				return err
-			}
+		results, err := r.reconcileDNSRecordHealthChecks(ctx, dnsRecord, config)
+		if err != nil {
+			return allResults, err
 		}
 
 		allResults = append(allResults, results...)
 
-		if err = r.updateDNSRecord(ctx, dnsRecordOriginal, dnsRecord); err != nil {
-			return err
+		if !equality.Semantic.DeepEqual(dnsRecordOriginal, dnsRecord) {
+			err = r.Client().Update(ctx, dnsRecord)
```

Comment:
When the reconciliation of the health checks was done in a different controller than the reconciliation of the DNS records it made sense that as a result of reconciling the health checks we would update the affected DNS records, but now that this is done in the same reconciliation loop I wonder if we should refactor it so that the `reconcileDNSRecords` method behaviour is affected by the reconciliation of the health checks.
The current approach still works but it feels weird that in the same reconciliation loop we are creating/updating the DNSRecords and then immediately (potentially) updating them again. Wdyt?

Reply:
Agreed, i went back and forth with this a few times. I wanted to have the dns records reconcile, return the records and then reconcile the health checks, and the last thing that happened was a single update to the record.

I went back on that for some reason, i think it was to keep the changes in this PR simpler initially, but i might go back and see if we can do it that way.

Reply:
As discussed, going to leave this as it is for now. We will revisit how we are doing this when we finalize what we are doing with health checks.